### PR TITLE
Example: Manual P2P

### PR DESCRIPTION
**Pull request by mouseless0x**
Example of peer communication using low level components from the [net crate](https://github.com/paradigmxyz/reth/tree/main/crates/net).

- Find peers using discv4
- Make a P2P handshake with a peer
- Make an ETH Wire handshake with a peer
- Listen in on peer's broadcasts (new txs & blocks)

### CODE REVIEW

File: examples/manual-p2p.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+#[tokio::main]
+async fn main() -> eyre::Result<()> {
+    // Setup discovery v4 protocol to find peers to talk to
+    let mut discv4_cfg = Discv4ConfigBuilder::default();
+    discv4_cfg.add_boot_nodes(mainnet_nodes());
+    discv4_cfg.lookup_interval(Duration::from_secs(1));
+
+    // Setup configs related to this 'node'
+    let our_key = rng_secret_key();
+    let our_peer_id = pk2id(&our_key.public_key(SECP256K1));
```
Comment (gakonst) [inline examples/manual-p2p.rs:35]: nit we should add a `let (sk, pk) = keygen()`-style function to make this easier

File: examples/manual-p2p.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    let our_enr = NodeRecord {
+        id: our_peer_id,
+        address: DEFAULT_DISCOVERY_ADDRESS.ip(),
+        tcp_port: DEFAULT_DISCOVERY_PORT,
+        udp_port: DEFAULT_DISCOVERY_PORT,
+    };
```
Comment (gakonst) [inline examples/manual-p2p.rs:41]: nit we should add a `fn new(...)` method that takes our ID and uses defaults

File: examples/manual-p2p.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    // Start discovery protocol
+    let discv4 =
+        Discv4::spawn(DEFAULT_DISCOVERY_ADDRESS, our_enr, our_key, discv4_cfg.build()).await?;
+    let mut discv4_stream = discv4.update_stream().await?;
+
+    while let Some(update) = discv4_stream.next().await {
+        tokio::spawn(async move {
+            if let reth_discv4::DiscoveryUpdate::Added(peer) = update {
+                // Boot nodes hard at work, lets not disturb them
+                if mainnet_nodes().contains(&peer) {
```
Comment (gakonst) [inline examples/manual-p2p.rs:52]: nit assign once to a variable instead of re-querying the function

**Review by gakonst — APPROVED** [review]
this is cool. no objections.  thank you for taking this on! curious if you have any api feedback improvement ideas?

find peer via discv4 -> p2p handshake -> eth handshake -> listen to msgs

maybe @mattsse has thoughts on API improvements so holding off on merging

File: examples/manual-p2p.rs (hunk 1 above)
Comment (Rjected) [reply examples/manual-p2p.rs:35]: `let (secret_key, pk) = SECP256K1.generate_keypair(&mut rng);` exists, so can probably just wrap that

Comment (gakonst) [reply examples/manual-p2p.rs:35]: (pk2id is different though)

File: examples/manual-p2p.rs
Comment (mattsse) [inline examples/manual-p2p.rs:1]: I'd like to move this to a standalone example crate instead

File: examples/manual-p2p.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        .build();
+
+    let status = Status { version: p2p_stream.shared_capability().version(), ..status };
+    let eth_unauthed = UnauthedEthStream::new(p2p_stream);
+    Ok(eth_unauthed.handshake(status, fork_filter).await?)
+}
+
+// Snoop by greedily capturing all peer's broadcasts
+// note: this node cannot handle request so will be disconnected by peer when challenged
+#[rustfmt::skip]
```
Comment (mattsse) [inline examples/manual-p2p.rs:118]: no skips

**Review by mattsse — CHANGES_REQUESTED** [review]

File: examples/manual-p2p.rs (hunk 2 above)
Comment (mouseless-eth) [reply examples/manual-p2p.rs]: Replaced this code with [NodeRecord::from_secret_key](https://paradigmxyz.github.io/reth/docs/reth_primitives/struct.NodeRecord.html#method.from_secret_key), which also solves the other nit by not having to derive the local_peer_id manually.

**Comment by mouseless-eth** [comment]
> this is cool. no objections. thank you for taking this on! curious if you have any api feedback improvement ideas?
> 
> find peer via discv4 -> p2p handshake -> eth handshake -> listen to msgs
> 
> maybe @mattsse has thoughts on API improvements so holding off on merging

Don't have any immediate feedback, setting up each layer was clear + intuitive enough

File: examples/manual-p2p/src/color_print_macros.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        println!("{}", format!($($arg)*).cyan().bold());
+    }};
+}
+
+#[macro_export]
+macro_rules! print_red_bold {
+    ($($arg:tt)*) => {{
+        println!("{}", format!($($arg)*).red().bold());
+    }};
+}
```
Comment (mattsse) [inline examples/manual-p2p/src/color_print_macros.rs:27]: imo we don't want to add new macros for examples, please remove

File: examples/manual-p2p/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+//!
+//! Run with
+//!
+//! ```not_rust
+//! cargo run -p manual-p2p
+//! ```
+
+use std::time::Duration;
+
+use colored::*;
```
Comment (mattsse) [inline examples/manual-p2p/src/main.rs:11]: I think we'll do just fine with regular prints, the messages are clear enough.

please use regular printlns

**Review by mattsse — CHANGES_REQUESTED** [review]
last style nits

File: examples/manual-p2p/src/main.rs (hunk 6 above)
Comment (mouseless-eth) [reply examples/manual-p2p/src/main.rs]: good point, the colored prints have been removed

**Review by mattsse — APPROVED** [review]
tysm for this

**Pull request merged**